packages/watcher/src/watcher.ts

```diff
@@ -0,0 +1,92 @@
+/* External Imports */
+import { ethers } from 'ethers-v4'
+
+interface Layer {
+  provider: any
+  messengerAddress: string
+}
+
+interface WatcherOptions {
+  l1: Layer
+  l2: Layer
+}
+
+export class Watcher {
+  public l1: Layer
+  public l2: Layer
+  public NUM_BLOCKS_TO_FETCH: number = 10_000_000
+
+  constructor(opts: WatcherOptions) {
+    this.l1 = opts.l1
+    this.l2 = opts.l2
+  }
+
+  public async getMessageHashesFromL1Tx(l1TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(true, l1TxHash)
+  }
+  public async getMessageHashesFromL2Tx(l2TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(false, l2TxHash)
+  }
+
+  public async getL1TransactionReceipt(l2ToL1MsgHash: string): Promise<any> {
+    return this._getLXTransactionReceipt(true, l2ToL1MsgHash)
+  }
+
+  public async getL2TransactionReceipt(l1ToL2MsgHash: string): Promise<any> {
+    return this._getLXTransactionReceipt(false, l1ToL2MsgHash)
+  }
+
+  private async _getMessageHashesFromTx(
+    isL1: boolean,
+    txHash: string
+  ): Promise<string[]> {
+    const layer = isL1 ? this.l1 : this.l2
+    const receipt = await layer.provider.getTransactionReceipt(txHash)
+    const filtered = receipt.logs.filter((log: any) => {
```

Comment:
You could do the combination of the filter + map in a single iteration using a loop.

```js
const data = []
for (const log of receipt.logs) {
  if (log.address === layer.messengerAddress && log.topics[0] === ethers.utils.id('SentMessage(bytes32)')) {
    data.push(log.data)
  }
}
return data
```